### PR TITLE
Fix UseManagedNtlm linker substitutions

### LINKED ISSUE #90898: NegotiateAuthentication larger code size and also brings in BigInteger
Afaik the following PR https://github.com/dotnet/runtime/pull/87930 is part of .NET 8.0 preview 7, it caused a 60kb size regression for us:

preview 6: https://github.com/npgsql/npgsql/actions/runs/5930686472
preview 7: https://github.com/npgsql/npgsql/actions/runs/5931742857

Mostly due to it bringing in BigInteger (and a managed NegotiateAuthentication impl?)

Not sure what we can or want to do here. 

/cc @filipnavara 

Screenshot 2023-08-22 at 00 04 06

### PR DESCRIPTION
**Pull request by filipnavara**
Depends on https://github.com/dotnet/sdk/pull/34903.

Tested with the following code:
```csharp
using System.Net.Security;
var negAuth = new NegotiateAuthentication(new NegotiateAuthenticationClientOptions { });
negAuth.GetOutgoingBlob("", out _);
```

It saves around 100Kb when this code is published with `dotnet publish -p:PublishAot=true`.

Fixes #90898

### CODE REVIEW

**Comment by filipnavara** [comment]
cc @sbomer @vitek-karas ... I find it odd that there was no other use of `--ignore-substitutions` in the code base. Maybe I am doing something wrong?

File: src/libraries/System.Net.Security/src/ILLink/ILLink.Substitutions.OSXLike.xml
Comment (teo-tsirpanis) [inline src/libraries/System.Net.Security/src/ILLink/ILLink.Substitutions.OSXLike.xml:1]: How about you call it `ILLink.Substitutions.Apple.xml`? `OSXLike` does not exist in any other file.

**Review by teo-tsirpanis — APPROVED** [review]

File: src/libraries/System.Net.Security/src/ILLink/ILLink.Substitutions.OSXLike.xml
Comment (filipnavara) [reply src/libraries/System.Net.Security/src/ILLink/ILLink.Substitutions.OSXLike.xml:1]: I have no strong preference on this, but Apple sounds good. There's at least one other case where `OSX` is used for both macOS and iOS.

**Comment by vitek-karas** [comment]
This feature switch is a bit weird currently:

* It's not documented in https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/feature-switches.md.
* There's no corresponding MSBuild property - even if "internal" only (underscore in front)

I would recommend:
* Adding an MSBuild property - if we don't want people to really change its value, keep it "internal", but this PR suggests that there are use cases for developers to change the value, so why not make it public. Without the property, the item group is the only way to modify this and that one is very cumbersome (it's easy to introduce duplicates and it's hard to test for the "current value", and os on)
* If we want this to be off by default on Linux, set the default value of the property to false on linux - in the SDK (not in the runtime's build, but in the app's build when this is used). That is assuming the default value everywhere else is true. It's OK if we leave the property undefined everywhere else - Linux is special in this case. It also seems we might want to default this to true for some other platforms (Mac, iOS, ...) - that would not change anything, but it would make the MSBuild property consistent with the actual runtime's behavior.
* Have the substitution file present on all platforms - by default it would not substitute anything (the feature switch value is undefined), and in library build it would not do anything either.

> Make the UseManagedNtlm property public because the linker and IL compiler doesn't support substitution of private properties.

That should not be the case - we have internal properties which are substituted - for example https://github.com/dotnet/runtime/blob/23b2ea3fd0a1be121ae7910496becb1dd0d978bf/src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs#L25 is substituted here https://github.com/dotnet/runtime/blob/23b2ea3fd0a1be121ae7910496becb1dd0d978bf/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml#L12

**Comment by filipnavara** [comment]
I am fine with introducing SDK property to control this. There are two reasons why I didn't do it at this point.

Firstly, I am not aware of any precedent where the default value of the property depends on the platform, and the default value was really a moving target (`false` on .NET 8 everywhere; `true` on .NET 9 on iOS/macOS; TBD: `true` on Linux Bionic). It feels really weird to have a default in SDK when we change it to fix platform-specific issues in dotnet/runtime. Moreover, the default has to match for non-linked and linked builds.

Secondly, I was considering a backport of minimal fix of this substitution to .NET 8. Doing that across several repos makes that prohibitively expensive fix.

> That should not be the case - we have internal properties which are substituted -

Let me rephrase that, it didn't work for `private` property when I tested it. ILLink didn't do the substitution during library build (thus I didn't realize that adding `featuredefault="true"` breaks things), and ILC didn't do it either. I had ILC under debugger and it didn't even enumerate the `private` method.

**Comment by jkotas** [comment]
> I had ILC under debugger and it didn't even enumerate the private method.

Where is this enumeration? I do not think ILC should be skipping private methods for substitutions.

**Comment by filipnavara** [comment]
> Where is this enumeration? I do not think ILC should be skipping private methods for substitutions.

https://github.com/dotnet/runtime/blob/0c4329bcb6f287d24d3b117af7199956d9d590ac/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/BodySubstitutionParser.cs#L59

**Comment by jkotas** [comment]
This does not have any filtering based on the visibility. It should be enumerating all methods.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets [hunk 1]
```diff
@@ -39,6 +39,10 @@ The .NET Foundation licenses this file to you under the MIT license.
     <EventSourceSupport Condition="$(EventSourceSupport) == ''">false</EventSourceSupport>
     <UseWindowsThreadPool Condition="'$(UseWindowsThreadPool)' == '' and '$(_targetOS)' == 'win'">true</UseWindowsThreadPool>
     <DynamicCodeSupport Condition="'$(DynamicCodeSupport)' == ''">false</DynamicCodeSupport>
+    <!-- Linux Bionic doesn't ship GSSAPI, so enable managed implementation -->
+    <_UseManagedNtlm Condition="'$(_UseManagedNtlm)' == '' and '$(_linuxLibcFlavor)' == 'bionic'">true</_UseManagedNtlm>
+    <!-- Trim managed NTLM on Linux when it's not explicitly requested -->
+    <_UseManagedNtlm Condition="'$(_UseManagedNtlm)' == '' and '$(_targetOS)' == 'linux'">false</_UseManagedNtlm>
```
Comment (eerhardt) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:45]: Why is this only for Native AOT apps? Wouldn't we want this default value set when just `PublishTrimmed=true`?

**Comment by filipnavara** [comment]
> This does not have any filtering based on the visibility. It should be enumerating all methods.

I tried to make a repro but it worked this time. Not sure what I did differently before, maybe the method got trimmed during library trimming and I forgot to reload in ILSpy. Sorry for the false alarm.

image
image

File: src/libraries/System.Net.Security/src/System/Net/NegotiateAuthenticationPal.Unix.cs [hunk 2]
```diff
@@ -19,7 +19,7 @@ namespace System.Net
 {
     internal partial class NegotiateAuthenticationPal
     {
-        private static bool UseManagedNtlm { get; } =
+        public static bool UseManagedNtlm { get; } =
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/src/System/Net/NegotiateAuthenticationPal.Unix.cs:22]: I don't think this should be flipped to `public`.

Are you adding an ILLink.Substitutions.xml entry for this new feature switch? Here's an example:

https://github.com/dotnet/runtime/blob/0c4329bcb6f287d24d3b117af7199956d9d590ac/src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml#L1-L15

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets (hunk 1 above)
Comment (filipnavara) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:45]: I am open to placing it on a better place but I didn't really find one. As stated above, it's unfortunate that this logic would have to be in sync between dotnet/sdk (linked build-time check) and dotnet/runtime (non-linked runtime check).

File: src/libraries/System.Net.Security/src/System/Net/NegotiateAuthenticationPal.Unix.cs (hunk 2 above)
Comment (filipnavara) [reply src/libraries/System.Net.Security/src/System/Net/NegotiateAuthenticationPal.Unix.cs:22]: I can revert it but it doesn't really matter since it's in non-public class. I had trouble getting `private` property to work correctly but it seems to have been an issue with my test setup since I cannot reproduce it anymore.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets (hunk 1 above)
Comment (filipnavara) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:45]: https://github.com/dotnet/runtime/blob/658bdd0376326f43cc8a6f316b2e01cacdcc688d/src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets#L34

Maybe this is the right place?

Comment (vitek-karas) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: > As stated above, it's unfortunate that this logic would have to be in sync between dotnet/sdk (linked build-time check) and dotnet/runtime (non-linked runtime check).

I view it as a hierarchy. The runtime check is obviously the ultimate decision maker - so for example, if the given feature doesn't exist on a given platform, runtime will simply not do it, no matter what. On platforms where it's optional, the responsibility is on the SDK to decide what the value should be - which is kind of the same as saying the app's developer should decide for a specific app.

The trimming process itself should not modify the value, it just "hardcodes" it into the IL (and while doing that removes the code which can't be used anymore after the hardcoding happens). Even without trimming SDK writes it into the app (`.runtimeconfig.json`) it just so happens that it's not completely hardcoded, but in reality nobody edits that JSON after build really.

Inside the SDK we have multiple considerations what to set the default value to - what we think makes sense for the developer. It would make sense to set it to false if the target platform doesn't support it, but not doing so doesn't really change much (other than consistency), unless we need it for size reasons, then we should do it. Other than that, we may for example decide on a different default for trimming/AOT. We typically do that if it's likely people don't need the feature and it saves a lot of size. An example of a similar approach is the globalization support, which is off for NativeAOT. But we can probably find others as well.

Comment (filipnavara) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: The `UseManagedNtlm` feature has the following considerations:
- It's supported on some Unix platforms only (ie. the code is hard-coded on Windows, Android, and tvOS)
- On the platforms where the switch has effect the user may override it to either `false` (only use GSSAPI, even if it's buggy on particular platform) or `true` (always use managed NTLM and SPNEGO implementations, use GSSAPI for Kerberos only).
- The default value is `false` on Linux/FreeBSD (Linux Bionic is odd exception, so let's ignore it for now), `true` on macOS / .NET 9+, `false` on macOS / .NET 8. This is implemented as fallback in the `UseManagedNtlm` property if the AppContext switch doesn't exist.
- If we are doing trimming and the default value is `false` according to the rules above we want to remove the managed NTLM/SPNEGO implementation to save space. It's not expected that the AppContext switch would be modified at runtime.

I updated the PR to set the `_UseManagedNtlm` property in Microsoft.NET.ILLink.targets. That seems to have effect both for `PublishTrimmed` and `PublishAot`. It's also part of the dotnet/runtime repository which lessens my concern about the defaults becoming out-of-sync.

Comment (filipnavara) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: To add a bit more context, the reasons why the user may want to change the value of the switch:
- On Linux the NTLM module for GSSAPI is optional, so setting `UseManagedNtlm=true` provides an alternative to installing GSS-NTLMSSP package. Many distributions also ship very old version of GSS-NTLMSSP, so this also ensure more consistent behaviour across all environments where the application is deployed.
- On macOS there are multiple known bugs in the NTLM implementation. This includes sending the credentials in slightly different way, which may result in failed authentication even for credentials that work on Linux/Windows, and a buffer overflows in the native code. Changing the default was deemed risky at the late .NET 8 stage, but we switched it for .NET 9. Thus on .NET 8 users may want to opt-in with `UseManagedNtlm=true`.
- For users that don't rely on NTLM but rely on Kerberos, setting `UseManagedNtlm=false` may provide size savings.

**Review by sbomer — APPROVED** [review]

File: src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets [hunk 3]
```diff
@@ -48,6 +48,10 @@ Copyright (c) .NET Foundation. All rights reserved.
     <_EnableConsumingManagedCodeFromNativeHosting Condition="'$(_EnableConsumingManagedCodeFromNativeHosting)' == ''">false</_EnableConsumingManagedCodeFromNativeHosting>
     <VerifyDependencyInjectionOpenGenericServiceTrimmability Condition="'$(VerifyDependencyInjectionOpenGenericServiceTrimmability)' == ''">true</VerifyDependencyInjectionOpenGenericServiceTrimmability>
     <JsonSerializerIsReflectionEnabledByDefault Condition="'$(JsonSerializerIsReflectionEnabledByDefault)' == ''">false</JsonSerializerIsReflectionEnabledByDefault>
+    <!-- Linux Bionic doesn't ship GSSAPI, so enable managed implementation -->
+    <_UseManagedNtlm Condition="'$(_UseManagedNtlm)' == '' and $(RuntimeIdentifier.StartsWith('linux-bionic'))">true</_UseManagedNtlm>
+    <!-- Trim managed NTLM on Linux when it's not explicitly requested -->
+    <_UseManagedNtlm Condition="'$(_UseManagedNtlm)' == '' and $(RuntimeIdentifier.StartsWith('linux'))">false</_UseManagedNtlm>
```
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets:54]: Presumably we could do it for FreeBSD too. I wanted to start small and get the approach right.

**Comment by eerhardt** [comment]
> This was because there are no other uses of featuredefault in substitutions (only in descriptors) - with the exception of nativeaot corelib, which I believe doesn't get trimmed in the runtime build. We so far only use substitution XMLs to either hard-code values during the library build (no feature or featuredefault attributes), or to add support for substituting when the application is trimmed.

See https://github.com/dotnet/linker/issues/3176. The lack of this is blocking https://github.com/dotnet/runtime/issues/80398 as well.

**Review by ivanpovazan — APPROVED** [review]
LGTM!

I do have one question. Would it make sense to enable trimming (via substitutions) even when `_UseManagedNtlm=true` ?

**Comment by filipnavara** [comment]
> Would it make sense to enable trimming (via substitutions) even when `_UseManagedNtlm=true` ?

Probably not. That code is still used for Kerberos regardless of the value of the switch.

**Review by wfurt — APPROVED** [review]
LGTM
